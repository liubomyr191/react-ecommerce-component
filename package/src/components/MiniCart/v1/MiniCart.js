import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { applyTheme, CustomPropTypes } from "../../../utils";

const Cart = styled.div`
  border: 1px solid ${applyTheme("color_black10")};
  max-width: 360px;
  overflow: hidden;
`;

const Items = styled.div`
  max-height: 420px;
  overflow-x: hidden;
  overflow-y: auto;
  padding-left: ${applyTheme("standardPadding")};
  padding-right: ${applyTheme("standardPadding")};
`;

const Footer = styled.div`
  border-top: 1px solid ${applyTheme("color_black10")};
  box-shadow: ${({ count }) => (count > 2 ? applyTheme("shadow_depth1") : "none")};
  padding: 0 1rem 1rem;
  position: relative;

  span {
    color: ${applyTheme("color_black30")};
    display: block;
    font-family: ${applyTheme("font_family")};
    font-size: ${applyTheme("font_size_small")};
    font-weight: normal;
    font-style: normal;
    font-stretch: normal;
    line-height: 1.25;
    letter-spacing: 0.3px;
    margin: 0.5rem 0 0;
    text-align: center;
  }
`;

class MiniCart extends Component {
  static propTypes = {
    /**
     * Cart data
     */
    cart: PropTypes.shape({
      /**
       * Cart checkout info
       */
      checkout: PropTypes.shape({
        /**
         * Checkout summary
         */
        summary: PropTypes.shape({
          /**
           * Checkout summary item total info
           */
          itemTotal: PropTypes.shape({
            /**
             * Checkout summary item total display amount
             */
            displayAmount: PropTypes.string
          }),
          /**
           * Checkout summary tax info
           */
          taxTotal: PropTypes.shape({
            /**
             * Checkout summary tax display amount
             */
            displayAmount: PropTypes.string
          })
        })
      }),
      /**
       * CartItem data. This is passed to CartItems, which may require some props.
       */
      items: PropTypes.arrayOf(PropTypes.object).isRequired
    }),
    /**
     * If you've set up a components context using @reactioncommerce/components-context
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      /**
       * Pass either the Reaction Button component or your own component that
       * accepts compatible props.
       */
      Button: CustomPropTypes.component.isRequired,
      /**
       * An element to show as the cart checkout button. If this isn't provided,
       * a button will be rendered using Button component.
       */
      cartCheckoutButton: PropTypes.node,
      /**
       * Pass either the Reaction CartItems component or your own component that
       * accepts compatible props.
       */
      CartItems: CustomPropTypes.component.isRequired,
      /**
       * Pass either the Reaction MiniCartSummary component or your own component that
       * accepts compatible props.
       */
      MiniCartSummary: CustomPropTypes.component.isRequired
    }),
    /**
     * On cart item quantity change handler
     */
    onChangeCartItemQuantity: PropTypes.func,
    /**
     * On default checkout button click. Not used if a custom button is supplied by `components.cartCheckoutButton`
     */
    onCheckoutButtonClick: PropTypes.func,
    /**
     * On remove item from cart handler
     */
    onRemoveItemFromCart: PropTypes.func
  };

  static defaultProps = {
    onChangeCartItemQuantity() {},
    onCheckoutButtonClick() {},
    onRemoveItemFromCart() {}
  };

  render() {
    const {
      cart: { checkout: { summary }, items },
      components: {
        Button,
        cartCheckoutButton,
        CartItems,
        MiniCartSummary,
        ...components
      },
      onCheckoutButtonClick,
      ...props
    } = this.props;
    return (
      <Cart>
        <Items>
          <CartItems items={items} components={components} {...props} isMiniCart />
        </Items>
        <Footer count={items.length}>
          <MiniCartSummary components={components} displaySubtotal={summary.itemTotal.displayAmount} />
          {cartCheckoutButton || <Button actionType="important" components={components} isFullWidth onClick={onCheckoutButtonClick}>Checkout</Button>}
          <span>Shipping and tax calculated in checkout</span>
        </Footer>
      </Cart>
    );
  }
}

export default withComponents(MiniCart);
