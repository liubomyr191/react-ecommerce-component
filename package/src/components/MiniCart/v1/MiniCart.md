### Overview

#### Usage

##### Default

```jsx
const checkout = {
  summary: {
    itemTotal: {
      displayAmount: "$25.00"
    },
    taxTotal: {
      displayAmount: "$2.50"
    }
  }
}

const items = [
{
  _id: "123",
  attributes: [{ label: "Color", value: "Red" }, { label: "Size", value: "Medium" }],
  compareAtPrice: {
    displayAmount: "$45.00"
  },
  currentQuantity: 3,
  imageURLs: {
    small: "//placehold.it/150",
    thumbnail: "//placehold.it/100"
  },
  isLowQuantity: true,
  price: {
    displayAmount: "$20.00"
  },
  productSlug: "/product-slug",
  productVendor: "Patagonia",
  title: "A Great Product",
  quantity: 2
},
{
  _id: "456",
  attributes: [{ label: "Color", value: "Black" }, { label: "Size", value: "10" }],
  currentQuantity: 500,
  imageURLs: {
    small: "//placehold.it/150",
    thumbnail: "//placehold.it/100"
  },
  isLowQuantity: false,
  price: {
    displayAmount: "$78.00"
  },
  productSlug: "/product-slug",
  productVendor: "Patagonia",
  title: "Another Great Product",
  quantity: 1
}];

const components = {
  CartCheckoutButtonComponent: () => <Button actionType="important" isFullWidth>Checkout</Button>,
  CartSummaryComponent: MiniCartSummary,
  CartItemsComponent: CartItems,
  CartItemComponent: CartItem,
  CartItemDetailComponent: CartItemDetail,
  CartItemStockWarningComponent: StockWarning,
  CartItemPriceComponent: Price,
  CartItemQuantityInputComponent: QuantityInput
};

<MiniCart cart={{ checkout, items }} components={components} />
```

##### Scrolling

```jsx
const checkout = {
  summary: {
    itemTotal: {
      displayAmount: "$25.00"
    },
    taxTotal: {
      displayAmount: "$2.50"
    }
  }
}

const items = [{
  _id: "123",
  attributes: [{ label: "Color", value: "Red" }, { label: "Size", value: "Medium" }],
  compareAtPrice: {
    displayAmount: "$45.00"
  },
  currentQuantity: 3,
  imageURLs: {
    small: "//placehold.it/150",
    thumbnail: "//placehold.it/100"
  },
  isLowQuantity: true,
  price: {
    displayAmount: "$20.00"
  },
  productSlug: "/product-slug",
  productVendor: "Patagonia",
  title: "A Great Product",
  quantity: 2
},
{
  _id: "456",
  attributes: [{ label: "Color", value: "Black" }, { label: "Size", value: "10" }],
  currentQuantity: 500,
  imageURLs: {
    small: "//placehold.it/150",
    thumbnail: "//placehold.it/100"
  },
  isLowQuantity: false,
  price: {
    displayAmount: "$78.00"
  },
  productSlug: "/product-slug",
  productVendor: "Nike",
  title: "Another Great Product",
  quantity: 1
},
{
  _id: "789",
  attributes: [{ label: "Color", value: "White" }, { label: "Size", value: "Medium" }],
  currentQuantity: 3,
  imageURLs: {
    small: "//placehold.it/150",
    thumbnail: "//placehold.it/100"
  },
  isLowQuantity: false,
  price: {
    displayAmount: "$20.00"
  },
  productSlug: "/product-slug",
  productVendor: "Everlane",
  title: "A Product",
  quantity: 2
},
{
  _id: "112",
  attributes: [{ label: "Color", value: "Black" }, { label: "Size", value: "10" }],
  currentQuantity: 500,
  imageURLs: {
    small: "//placehold.it/150",
    thumbnail: "//placehold.it/100"
  },
  isLowQuantity: false,
  price: {
    displayAmount: "$150.00"
  },
  productSlug: "/product-slug",
   productVendor: "Greats",
  title: "Another Product",
  quantity: 1
}];

const components = {
  CartCheckoutButtonComponent: () => <Button actionType="important" isFullWidth>Checkout</Button>,
  CartSummaryComponent: MiniCartSummary,
  CartItemsComponent: CartItems,
  CartItemComponent: CartItem,
  CartItemDetailComponent: CartItemDetail,
  CartItemStockWarningComponent: StockWarning,
  CartItemPriceComponent: Price,
  CartItemQuantityInputComponent: QuantityInput
};

<MiniCart cart={{ checkout, items }} components={components} />
```
