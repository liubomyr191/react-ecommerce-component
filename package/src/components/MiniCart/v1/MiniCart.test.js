import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import MiniCart from "./MiniCart";

test("basic snapshot", () => {
  const component = renderer.create(<MiniCart />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
