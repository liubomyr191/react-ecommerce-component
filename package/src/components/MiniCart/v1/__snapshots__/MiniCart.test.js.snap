// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot with props 1`] = `
.c0 {
  border: 1px solid #e6e6e6;
  max-width: 360px;
  overflow: hidden;
}

.c1 {
  max-height: 420px;
  overflow-x: hidden;
  overflow-y: auto;
  padding-left: 10px;
  padding-right: 10px;
}

.c2 {
  border-top: 1px solid #e6e6e6;
  box-shadow: none;
  padding: 0 1rem 1rem;
  position: relative;
}

.c2 span {
  color: #b3b3b3;
  display: block;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  font-size: 14px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.25;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  margin: 0.5rem 0 0;
  text-align: center;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    CartItems(undefined)
  </div>
  <div
    className="c2"
  >
    MiniCartSummary(undefined)
    Button(undefined)
    <span>
      Shipping and tax calculated in checkout
    </span>
  </div>
</div>
`;
